### ISSUE TITLE
Tpetra: FECrsMatrix ctor sets active=owned+shared, and fillState=closed, regardless of graph status.

### ISSUE THREAD

**Issue by bartgol** [trilinos/Trilinos#9638]
## Enhancement
@trilinos/tpetra 

When building a FECrsMatrix from a FE graph, the active matrix is set to be the owned+shared one, and the fillState is set to closed. I find this confusing and possibly wrong, for two reasons:

1. fillState should match the active matrix. So if active=owned+shared, fillState should be "open", otherwise one between "modify" and "closed". 
2. if one builds a FE mat from a pre-assembled graph, they might not need an assembly phase. Creating the matrix in owned+shared mode, even if the input graph is closed, forces to do a pointless global assembly.

Additionally, the matrix might be filled inside a routine, which likely will start with `beginAssembly` or `beginModify` (depending on the type of filling). The routine should work the same, regardless of whether the matrix was newly created, or simply reused.

To sum it up, I would expect the FE matrix to satisfy the following:

- if input graph in ctor is closed, active=owned, and fillState=closed
- if input graph in ctor is open, active=owned+shared, and fillState=open
- at any given time, if active=owned+shared, fillState=open, and if active=owned, fillState=closed or fillState=modify

**Comment by kddevin** [comment]
@tjfulle @csiefer2

**Comment by tjfulle** [comment]
Hey @bartgol, thank you for the issue!  I'd like to understand the use case a little better.  I'm a little confused to the last bullet points you listed:

> 
> - if input graph in ctor is closed, active=owned, and fillState=closed
> - if input graph in ctor is open, active=owned+shared, and fillState=open
> - at any given time, if active=owned+shared, fillState=open, and if active=owned, fillState=closed or fillState=modify
>

to the first and second points, the input graph should always be closed.  I suppose it is not a hard requirement because we check for `graph->fillComplete()` and not the `FillState`, but that is the intent.  That being the case, if the graph is closed, we can't really assume what the user wants to do with the matrix.  So, we initialize it with `fillState=closed` and require the user to explicitly declare intent with `beginAssembly` and `beginModify`.  To the last point, a design decision of `FECrsMatrix` was to make requests RE the `fillState` explicit.  But the call to `beginAssembly` and/or `beginModify` shouldn't consume too many resources, all they do is check consistency with the `fillState` and which (portion of the) matrix is active.  The only communication occurs during `endAssembly`.

Am I missing something from the Albany use case?

**Comment by bartgol** [comment]
Well, in our case we recycle the graph to build several matrices. We build the graph once, we close it (meaning, we 'endAssembly' it), and then build several matrices later on. The graph builder is itself a self contained utility: given a connectivity structure (of some sort) between dofs in the range and domain discretizations, it builds a graph.

I just so happens that in some circumstances, the graph built is that of a projector, with one entry per row, and no off-processor entries. In this case, I'd like to do the following (in pseudo code):

```
auto graph = graph_builder.build(...);
FECrsMatrix mat (graph);
mat.setAllToScalar(1.0);
```
Now, I saw that #9660 will simplify life in terms of not requiring a resume fill in order to call `setAllToScalar`, and that's great. However, it seems that `mat` is created with "ownedPlusShared" as the active matrix, forcing me to call endAssembly for nothing. If, instead, `mat` were to be created with "owned" as the active matrix, i'd be good to go, with no further comm needed.

As you pointed out:
> [...] a design decision of FECrsMatrix was to make requests RE the fillState explicit.

That's my point: if a graph is not still "open" (I don't even know if one can create a FEMat from a non-filled graph, but I assume so), then it makes sense to create the mat with active=ownedPlusShared. But if the graph is closed, then the user should make the request of beginning the assembly phase, not Tpetra. Put it another way: since opening is "free", as you mentioned, Tpetra should create the matrix in "owned" state when possible, since opening it is a no-op; but creating it in owned+shared mode when not strictly necessary, may require the user to call endAssembly even if not needed.

**Comment by tjfulle** [comment]
@bartgol - the graph must have be fill complete before constructing the corresponding matrix.  Since it is always fill complete, the matrix has no way to ascertain from the graph that all future operations will be strictly local (ie, no `endAssembly` required).  I suppose an optional flag could be sent to the constructor?  @csiefer2, any thoughts?

**Comment by bartgol** [comment]
Well, since "opening" the matrix is free, FECrsMatrix should probably create the matrix "closed", and let the user open it if global operations are needed, no? I mean:

- Mat created open: good for global ops needed, bad for no-global ops needed (cause endAssembly is necessary)
- Mat created closed: good for no-global ops needed, fair for global-ops needed (calling beginAssembly is cheap)

**Comment by tjfulle** [comment]
@bartgol - are you good with the solution in #9811?  If so, I'll close this issue.

**Comment by bartgol** [comment]
Yes, that's fine. Thank you!
